Draft upgrade procedure to single service API ML deployment (#4983)
zowe/docs-site: docs/user-guide/api-mediation/upgrade-to-single-service-apiml-deployment.md

:::

## System Requirements & Prerequisites
Before transitioning to the single-service model, verify that your z/OS environment meets the minimum standards for Zowe v3:

* **Java**  
Must be upgraded to Java 17 or Java 21. Ensure the `java.home` parameter in your zowe.yaml points to the correct home location for one of these versions.

* **Node.js**  
Upgrading to v20 is recommended. Update the `node.home` parameter in your zowe.yaml accordingly.

* **Authentication**  
SAF (System Authorization Facility) is the recommended authentication provider for enterprise security. Enabling SAF ensures that Zowe uses your existing security manager (ESM) for identity and access management.

* **z/OSMF**  
If you are using z/OSMF as the authentication provider, the following versions are supported: V2R5, V3R1, or V3R2.

* **JWT Support**  
Zowe v3 uses JSON Web Token (JWT) as the primary mechanism for session management and Single Sign-On (SSO), while SAF is the recommended provider for identity and resource authorization. To enable JWT support, you must have a signing certificate (stored in a PKCS12 keystore or SAF keyring).   

>   * Ensure the Gateway has access to a private key for signing, and that the public certificate is distributed to any downstream services requiring authentication.

Author: Please confirm that these bulleted statements are accurate.

Reviewer: the certificateAlias is not correct I believe. 

As for the part before, it may be misleading a bit. 
API Mediation Layer supports various method of authentication from the client side. One of the options is a JWT token issues by API ML (the actual issuance is delegated to the z/OSMF in case of z/OSMF as authentication provider)

What may be more in line with reality:
- API Mediation Layer needs to validate the user's identity and authenticate it. There are two primary options to do so - z/OSMF Rest APIs, SAF Direct APIs. 
- One of the tools to integrate with API ML SSO is a JWT token that is produced by calling login endpoint. This token is also used as primary method of providing authentication to downstream services running on z/OS
- There are alternative methods, that does not necessarily use the JWT token on the client side - Client Certificates (ideal for service to service scenarios), Personal Access Tokens (For usage within specific Dev Tools or as an alternative to the client certificates for long living but scoped tokens), OIDC (Ideal for day to day interactive work)

Author: Please see the refactored section.

>   * Verify that `zowe.components.apiml.security.token.certificateAlias` in your zowe.yaml correctly points to your signing certificate.

Author: Should this be `zowe.configs`?

Reviewer 2: there is no such property as "zowe.components.apiml.security.token.certificateAlias"

Author: Could you please advise. Perhaps pointing me to the yaml would help.

Reviewer 2: I don't know what is meant by this point. Which signing certificate do you mean? API ML JWT is signed by the private key of the server certificate, and there is no option to change it. Is this some other signing certificate?

Author: Please see the refactored section.

Reviewer 2: now it looks good to me, thanks for the update

:::info Required role: system administrator
:::

Single-service deployment reduces the Zowe footprint by running the Gateway, Discovery Service, API Catalog, Caching Service, and ZAAS within a single address space.

:::note
For comprehensive information about single-service deployment, performance improvements, and internal service consolidation, see [Enabling single-service deployment of API Mediation Layer](../api-mediation/api-mediation-modulith.md).
:::

## System requirements and prerequisites
Before transitioning to the single-service model, verify that your z/OS environment meets the minimum standards for Zowe v3:

* **Java**  
Must be upgraded to Java 17 or Java 21. Ensure the `java.home` parameter in your zowe.yaml points to the correct home location for one of these versions.

* **Node.js**  
Upgrading to v20 is recommended. Update the `node.home` parameter in your zowe.yaml accordingly.

> * **Authentication and Identity Validation**  

Reviewer: Wouldn't it be better to have this and Ensure JWT Support as subsections instead of just nested bullet points?

Author: agreed. Changed to sections

SAF (System Authorization Facility) is the recommended authentication provider for enterprise security. Enabling SAF ensures that Zowe uses your existing security manager (ESM) for identity and access management.

* **z/OSMF**  
If you are using z/OSMF as the authentication provider, the following versions are supported: V2R5, V3R1, or V3R2.

* **JWT Support**  
Zowe v3 uses JWT as the primary mechanism for session management and Single Sign-On (SSO), while SAF is the recommended provider for identity and resource authorization. 

:::tip
Ensure that all Zowe address spaces are stopped before modifying configuration files.
:::

## Transitioning to Single-Service Mode
To migrate from the multi-service deployment to the single-service deployment, apply these changes to your zowe.yaml file:

Enable single-service deployment.

> Set `components.apiml.enabled` to `true` under the `components.gateway` configuration:

Reviewer 2: ```suggestion
Set `components.apiml.enabled` to `true`:
```

:::note
All standalone components from the multi-service API ML deployment including the Discovery Service, API Catalog, and Caching Service are internal to the Gateway address space in the single-service deployent and are disabled when `apiml.enabled` is configured.
:::

>   ```yaml

Reviewer 2: ```suggestion
  components:
      apiml:
        enabled: true
```

> :::tip

Reviewer: Is it possible to use ::caution instead of ::tip as it is quite important to do that before the upgrade?

Author: changed to caution.

  components:
    gateway:
      apiml:
        enabled: true
  ```

:::note Alternative authentication with z/OSMF

> While SAF is the recommended authentication provider, it is possible to use z/OSMF  authentication with a Java Web Token (JWT) if you cannot use SAF.

Reviewer 2: ```suggestion
While SAF is the recommended authentication provider, it is possible to use z/OSMF authentication with a JSON Web Token (JWT) if you prefer not to use SAF.
```

>   components:

Reviewer: I am sorry for the nitpicking, but with the amount of issue we saw with yaml, can we make sure that the indentation is consistently sized? 

E.g. if we use 2 for the first depth, then for second we use 4 and then 6 for the third?

  components:
    apiml: 
      enabled: true

That is how it would look like for this one. If that is not enough, we should use 4 but again across the whole block.

Author: Fixed. I used the indentation as indicated in example-zowe.yaml

      apiml:
        enabled: true
  ```

:::note

> All standalone components from the multi-service API ML deployment including the Discovery Service, API Catalog, and Caching Service are internal to the Gateway address space in the single-service deployent and are disabled when `apiml.enabled` is configured.

Reviewer: deployment not deployent

Reviewer: We probably also should use the whole path components.apiml.enabled instead of just apiml.enabled

Author: updated

1. To use z/OSMF as authentication provider, set `auth.provider` to `zosmf`, and `jwtAutoconfiguration` to `jwt`:
  
```yaml
  components:
    gateway:
      apiml:
        security:
          auth:
            provider: zosmf
            zosmf:
              jwtAutoconfiguration: jwt
              serviceId: ibmzosmf
```

2. Disable SAF resource checking: 

```yaml
  components:
    gateway:
      apiml:
        security:
          authorization:
            saf:
              enabled: false              

> ```

Reviewer 2: this is not relevant for the zOSMF provider. I would rather delete it

> While SAF is the recommended authentication provider, it is possible to use z/OSMF authentication with a JSON Web Token (JWT) if you prefer not to use SAF.

Reviewer: What we still may want to clear, is that LTPA is still supported even though it is not generally recommended.

Author: updated with this inclusion.

For details about using JWT and the token lifecycle, see [Authenticating with a JWT token](../authenticating-with-jwt-token.md).
  
</details>

:::

## Refreshing infrastructure and cleanup

Update the procedure library by running the command `zwe init stc` (or use JCL `ZWEISTC`). This update ensures your `ZWESLSTC` and `ZWESISTC` members are updated for v3.

:::note
If using Zowe Cross-Memory Server (ZIS), ensure that you update APF Authorization. 
Re-run `zwe init apfauth` to ensure the ZIS load libraries are authorized.
:::

## Restarting Zowe

After completing the infrastructure refresh and mandatory cleanup, you can bring Zowe back online.

:::note
If you are using ZIS, start this component first. This server provides the authorized services required by Zowe. 
```
/S ZWESISTC 
```
For details about ZIS configuration, see [Configuring the Zowe cross memory server (ZIS)](../configure-xmem-server.md).
:::

Start the main Zowe task. 

>     ```

Reviewer: Is the indentation here, intentionall?

Author: reformatted

# Upgrading to single-service API ML deployment

Enable the single-service deployment of API ML for an existing multi-service API ML deployment. This procedure is specifically designed for users upgrading from **Zowe v2.18** (the final v2 release) or **Zowe v3.3.x** multi-service deployments to the consolidated single-service deployment model.

:::info Required role: system administrator
:::

> Single-service deployment reduces the Zowe footprint by running the Gateway, Discovery, and API Catalog, the Caching service and ZAAS components within a single address space.

Reviewer 3: Maybe better as:

Single-service deployment reduces the Zowe footprint by running the Gateway, Discovery, API Catalog, Caching service and ZAAS components within a single address space.

# Upgrading to single-service API ML deployment

Enable the single-service deployment of API ML for an existing multi-service API ML deployment. This procedure is specifically designed for users upgrading from **Zowe v2.18** (the final v2 release) or **Zowe v3.3.x** multi-service deployments to the consolidated single-service deployment model.

:::info Required role: system administrator
:::

> Single-service deployment reduces the Zowe footprint by running the Gateway, Discovery, and API Catalog, the Caching service, and ZAAS components within a single address space.

Reviewer: Proposed to replace ... running the Gateway ...

...running the Gateway, Discovery Service, API Catalog, Caching Service, and ZAAS within a single address space.

Author: Changes have been applied.

:::note
For comprehensive information about single-service deployment, performance improvements, and internal service consolidation, see [Enabling single-service deployment of API Mediation Layer](../api-mediation/api-mediation-modulith.md).
:::

> ## System Requirements & Prerequisites

Reviewer: Should we use and instead of ampersand?

Author: fixed

Zowe v3 utilizes JSON Web Tokens (JWT) as the primary mechanism for Single Sign-On (SSO). JWT provides the authentication context to downstream services.

  * **When using SAF**  
  The Gateway signs the JWT directly using the private key of the server certificate. Note that API ML exclusively uses the server's identity certificate for this purpose. Ensure the Gateway has access to a private key (stored in a PKCS12 keystore or SAF keyring) for signing.

    Set `components.gateway.apiml.security.auth.provider: saf`. The certificate that is used for signing is found in `zowe.certificate.keystore.file`. This configuration enables API ML to validate identities using SAF Direct APIs.

  * **When using z/OSMF**  
  Token issuance is delegated to z/OSMF, wherein the JWT is produced by the z/OSMF authentication provider.

:::tip
Ensure that all Zowe address spaces are stopped before modifying configuration files.
:::

## Transitioning to Single-Service Mode
To migrate from the multi-service deployment to the single-service deployment, apply these changes to your zowe.yaml file:

Enable single-service deployment.
Set `components.apiml.enabled` to `true`:

:::note

> All standalone components from the multi-service API ML deployment including the Discovery Service, API Catalog, and Caching Service are internal to the Gateway address space in the single-service deployent and are disabled when `apiml.enabled` is configured.

Reviewer: single-service deployment instead of deployent

Author: I don't understand what you are referring to. All mentions of deployment specify if this is single-service or multi-service.

Reviewer: There is a typo in the paragraph

:::
  ```
  components:
      apiml:
        enabled: true

>   components:

Reviewer: Actually I believe that the second components section must be removed.

Reviewer: Would it make sense to have this section before the note?

Author: Moved yaml before note.

    gateway:
      apiml:
        enabled: true
  ```

:::note Alternative authentication with z/OSMF
While SAF is the recommended authentication provider, it is possible to use z/OSMF authentication with a JSON Web Token (JWT) if you prefer not to use SAF.

<details>

<summary>Click here for details about using z/OSMF as your authentication provider.</summary>

**z/OSMF-Based Authentication (Not Recommended)** 

To use z/OSMF as authentication provider, set `auth.provider` to `zosmf`, and `jwtAutoconfiguration` to `jwt`:
  
```yaml
  components:
    gateway:
      apiml:
        security:
          auth:
            provider: zosmf
            zosmf:
              jwtAutoconfiguration: jwt
              serviceId: ibmzosmf
```

For details about using JWT and the token lifecycle, see [Authenticating with a JWT token](../authenticating-with-jwt-token.md).
  
</details>

:::

> ## Refreshing Infrastructure & Cleanup

Reviewer: Same as above is ampersand instead of and intentional?

Author: fixed

# Upgrading to single-service API ML deployment

Enable the single-service deployment of API ML for an existing multi-service API ML deployment. This procedure is specifically designed for users upgrading from **Zowe v2.18** (the final v2 release) or **Zowe v3.3.x** multi-service deployments to the consolidated single-service deployment model.

:::info Required role: system administrator:::

> Single-service deployment reduces the Zowe footprint by running the Gateway, Discovery, and API Catalog components within a single address space.

Reviewer: And also Caching Service and ZAAS

Author: addressed

:::note
For comprehensive information about single-service deployment, performance improvements, and internal service consolidation, see [Enabling Single-Service deployment of API Mediation Layer](../api-mediation/api-mediation-modulith.md).
:::

## System Requirements & Prerequisites
Before transitioning to the single-service model, verify that your z/OS environment meets the minimum standards for Zowe v3:

* **Java:**  
Must be upgraded to **Java 17** or **Java 21**. Update the `java.home` parameter in your `zowe.yaml`.
* **Node.js:**  

> Must be upgraded to **v18 or v20**. Update `node.home`.

Reviewer: Let's focus on at least v20, but technically speaking this is not pre-req for existing installation upgrades

Author: updated

* **z/OSMF:**  

> Requires **V2R5** or **V3R1**. 

Reviewer: V3R2 works as well. 
But it is possible to run without z/OSMF

Author: please see the update

> * **JWT Support:**  

Reviewer: The actual recommendation is to move to the SAF as authentication provider and thus keeping z/OSMF only for access and not authentication.

Author: SAF is now the primary option with JWT included as an alternative (see the note).

Highly recommended. If JWT support is not available, set `jwtAutoconfiguration` to `ltpa`.
:::tip
Ensure that all Zowe address spaces are stopped before modifying configuration files.
:::

## Enabling the ZAAS Component

> In Zowe v3, the **Zowe Authentication and Authorization Service (ZAAS)** is a mandatory standalone component for API ML. In previous versions, this functionality was embedded within the Gateway.

Reviewer: That is true only if you have multi-service deployment, in single-service it is part of single-service

Author: As this is part of enabling the single-service, this section has been removed.

Ensure that your `zowe.yaml` contains the following section with the specified parameter values:

```yaml
components:
  zaas:
    enabled: true
    port: 7558
    debug: false
```
:::note
Port `7558` was previously used by the `jobs-api`. Ensure the `jobs-api` component is removed to avoid a port conflict with ZAAS.
:::

## Transitioning to Single-Service Mode
To migrate from the multi-service deployment to the single-service deployment, apply these changes to your zowe.yaml file:

1. Enable single-service deployment.

> Set `deploymentMode` to single-service under the gateway configuration:

Reviewer: I believe the correct answer is:
components:
   apiml:
       enabled: true

```yaml
components:
  gateway:
    apiml:
      deploymentMode: single-service
```

2. Disable standalone components.
Because the Discovery Service, API Catalog, and Caching Service are now internal to the Gateway address space, these components must be disabled to prevent port conflicts and unnecessary address spaces:

```yaml

> components:

Reviewer: Also gateway I believe should be disabled, but I also believe that just setting the components.apiml.enabled to true automatically disables them.

Author: Confimed by Andrej. This has been removed.

  discovery:
    enabled: false
  api-catalog:
    enabled: false
  caching-service:
    enabled: false
```
3. Remove deprecated components. 
To prevent "component not found" errors, remove the configuration blocks for these services that no longer exist in v3:

* `metrics-service`
* `jobs-api`
* `files-api`

## Authentication & Gateway Updates
Update the Gateway’s security configuration to reflect the service ID changes in v3:

1. Update your service ID to `ibmzosmf`.

> 2. Set `jwtAutoconfiguration` to `jwt` (preferred) or `ltpa`.

Reviewer: I would recommend the move to the SAF

Author: SAF is now the recommended authentication provider

```yaml
components:
  gateway:
    apiml:
      security:
        auth:
          zosmf:
            jwtAutoconfiguration: jwt
            serviceId: ibmzosmf
```

## Refreshing Infrastructure & Cleanup

1. Update PROCLIB. 
Run the command `zwe init stc` (or use JCL `ZWEISTC`) to ensure your `ZWESLSTC` and `ZWESISTC` members are updated for v3.

2. Update APF Authorization. 

> Re-run `zwe init apfauth` to ensure the ZIS load libraries are authorized.

Reviewer: This applies only if ZIS is actually used.

Author: Updated as a note

## Restarting Zowe

After completing the infrastructure refresh and mandatory cleanup, you can bring Zowe back online.

> 1. Start the Zowe Cross-Memory Server (ZIS).

Reviewer: Relevant only when ZSS is used, it is not required.

Author: See note

If you stopped the Cross-Memory server during the upgrade, start ZIS first. This server provides the authorized services required by Zowe.

    ```
    /S ZWESISTC 
    ```

1. Start the main Zowe task. The launcher now initializes the consolidated single-service instead of multiple separate address spaces for the API Mediation Layer.
    ```
    /S ZWESLSTC
    ```
2. Verify the consolidated address spaces.
In a single-service deployment, your SDSF (or equivalent) should show a simplified list of address spaces. Specifically, you should see:

* `ZWESL` (or similar): The main launcher.
* `GW` suffix: The consolidated Gateway (now hosting Discovery and Catalog).

> * `AZ` suffix: The new ZAAS (Zowe Authentication and Authorization Service).

Reviewer: Will not be there.

Author: removed

:::info Required role: system administrator
:::

Single-service deployment reduces the Zowe footprint by running the Gateway, Discovery, and API Catalog, the Caching service and ZAAS components within a single address space.

:::note
For comprehensive information about single-service deployment, performance improvements, and internal service consolidation, see [Enabling Single-Service deployment of API Mediation Layer](../api-mediation/api-mediation-modulith.md).
:::

## System Requirements & Prerequisites
Before transitioning to the single-service model, verify that your z/OS environment meets the minimum standards for Zowe v3:

* **Java**  
Must be upgraded to Java 17 or Java 21. Ensure the `java.home` parameter in your zowe.yaml points to the correct home location for one of these versions.

* **Node.js**  
Upgrading to v20 is recommended. Update the `node.home` parameter in your zowe.yaml accordingly.

* **Authentication**  
SAF (System Authorization Facility) is the recommended authentication provider for enterprise security. Enabling SAF ensures that Zowe uses your existing security manager (ESM) for identity and access management.

* **z/OSMF**  

> If you are using z/OSMF to communicate to SAF, the following versions are supported: V2R5, V3R1, or V3R2.

Reviewer 2: ```suggestion
If you are using z/OSMF as the authentication provider, the following versions are supported: V2R5, V3R1, or V3R2.
```

* **JWT Support**  

> Zowe v3 uses JWT as the primary mechanism for session management and Single Sign-On (SSO), while SAF is the recommended provider for identity and resource authorization. If JWT support is not yet configured or available in your z/OSMF environment, set `jwtAutoconfiguration` to `ltpa` to maintain SAF-backed authentication using legacy session tokens.

Reviewer 2: if the Authentication recommends SAF, then I don't understand this point. "jwtAutoconfiguration" is only for zOSMF and has no impact on SAF auth provider. I think it would be better to remove this point

Author: addressed

:::tip
Ensure that all Zowe address spaces are stopped before modifying configuration files.
:::

## Transitioning to Single-Service Mode
To migrate from the multi-service deployment to the single-service deployment, apply these changes to your zowe.yaml file:

Enable single-service deployment.
Set `deploymentMode` to single-service under the gateway configuration:

  ```yaml
  components:
    gateway:
      apiml:
        deploymentMode: single-service
  ```

:::note
All standalone components from the multi-service API ML deployment including the Discovery Service, API Catalog, and Caching Service are internal to the Gateway address space in the single-service deployent and are disabled when `apiml.enabled` is configured.
:::

> ## Configuring authentication 

Reviewer 2: what is the goal of this section?

Author: addressed

1. Set `jwtAutoconfiguration` to `ltpa` to use SAF-backed tokens for z/OSMF.

  ```yaml
  components:
    gateway:
      apiml:
        security:
          auth:
            zosmf:
              jwtAutoconfiguration: ltpa
              serviceId: ibmzosmf
  ```
API ML now uses z/OSMF as its identity validator.

2. Enable SAF authorization.
   
  ```yaml
  components:
    gateway:
      apiml:
        security:
          authorization:
            saf:
              enabled: true
  ```
API ML now checks specific SAF resource profiles before a request is allowed to pass through the Gateway.

:::note
While SAF is the recommended authentication provider, it is possible to use JWT-based authentication if you cannot use SAF.

<details>
<summary>Click here for details about using JWT as your authentication provider.</summary>

**JWT-Based Authentication (Not Recommended)** 

While SAF is the recommended provider for enterprise production environments, you may choose to use **JSON Web Token (JWT)** authentication for specific use cases. 

1. Set `jwtAutoconfiguration` to `jwt`:
   

> ```yaml
>   components:
>     gateway:
>       apiml:
>         security:
>           auth:
>             zosmf:
>               jwtAutoconfiguration: jwt
>               serviceId: ibmzosmf

Reviewer 2: ```suggestion
In case you want to continue using a zOSMF as authentication provider, you need to set:
  components:
    gateway:
      apiml:
        security:
          auth:
            provider: zosmf
            zosmf:
              jwtAutoconfiguration: jwt
              serviceId: ibmzosmf
```

```

2. Disable SAF resource checking: 

```yaml
  components:
    gateway:
      apiml:
        security:
          authorization:
            saf:
              enabled: false              
```

For details about using JWT and the token lifecycle, see [Authenticating with a JWT token](../authenticating-with-jwt-token.md).
  
</details>

:::

## Refreshing Infrastructure & Cleanup

Update PROCLIB.  
Run the command `zwe init stc` (or use JCL `ZWEISTC`) to ensure your `ZWESLSTC` and `ZWESISTC` members are updated for v3.

:::note
If using Zowe Cross-Memory Server (ZIS), ensure that you update APF Authorization. 
Re-run `zwe init apfauth` to ensure the ZIS load libraries are authorized.
:::

## Restarting Zowe

After completing the infrastructure refresh and mandatory cleanup, you can bring Zowe back online.

:::note
If you are using ZIS, start this component first. This server provides the authorized services required by Zowe. 
```
/S ZWESISTC 
```
For details about ZIS configuration, see [Configuring the Zowe cross memory server (ZIS)](../configure-xmem-server.md).
:::

Start the main Zowe task. 
    ```
    /S ZWESLSTC
    ```

> The launcher now initializes the consolidated single-service instead of multiple separate address spaces for the API Mediation Layer.

Reviewer 2: ```suggestion
The launcher now initializes the API Mediation Layer in a single address space instead of multiple address spaces.
```
